## Pandas

 **(a)** Import the `pandas` module under the name `pd`.

In [4]:
import pandas as pd

## 1. Format of a DataFrame

 A `DataFrame` is in the form of a **matrix** whose rows and columns each have an **index**. Typically, columns are indexed by name and rows by unique identifiers.

 A `DataFrame` is used to store a **database**. The different **entries** in the database (individuals, animals, objects, etc.) are the different **lines** and their **features** are the different **columns**:


 |  |  Name   |  Gender |  Height |Age|
 |--|--------|-------|--------|---|
 |**0** |  Robert|   M   |   174  |23 |
 |**1** |  Mark  |   M   |   182  |40 |
 |**2** |  Aline |   F   |   169  |56 |


 * The `DataFrame` above groups together information on **3 individuals**: the `DataFrame` therefore has **3 lines**.


 * For each of these individuals, there are **4 variables** (name, gender, height and age) : therefore, the `DataFrame` has **4 columns**.



 The column containing **the numbering of the lines** is called the **index** and is not managed in the same way as other columns of the `DataFrame`.

 The index can be set by default (will follow the row numbering), defined with one (or several) of the columns of the `DataFrame` or even defined with a list that we specify.

 <span style="color:#09b038; text-decoration : underline"> Example: </span> Default indexing (line numbering), you don't have to specify anything :

 |  |  Name   |  Gender |  Height |Age|
 |--|--------|-------|--------|---|
 |**0** |  Robert|   M   |   174  |23 |
 |**1** |  Mark  |   M   |   182  |40 |
 |**2** |  Aline |   F   |   169  |56 |

 <span style="color:#09b038; text-decoration : underline"> Example: </span> Indexing by the column `'Name'`:

 |        |  Gender |  Height |Age|
 |--------|-------|--------|---|
 |**Robert** |  M   |   174  |23 |
 |**Mark** |  M   |   182  |40 |
 |**Aline** | F   |   169  |56 |

 <span style="color:#09b038; text-decoration : underline"> Example: </span> Indexing by the list `['person_1', 'person_2', 'person_3']`:

 |  |  Name   |  Gender |  Height |Age|
 |--|--------|-------|--------|---|
 |**person_1** |  Robert|   M   |   174  |23 |
 |**person_2** |  Mark  |   M   |   182  |40 |
 |**person_3** |  Aline |   F   |   169  |56 |

 We will detail later how to define the index when creating a `DataFrame`.


 The `DataFrame` class has several advantages over a` Numpy` array:

 * Visually, a `DataFrame` is much more **readable** thanks to more explicit column and row indexing.


 * Within the same column the elements are of the same type but from one column to another, the **type of the elements may vary**, which is not the case of `Numpy` arrays which only support data of the same type.


 * The `DataFrame` class contains more methods for handling and preprocessing databases, while `NumPy` specializes instead in optimized computation.


## 2. Creation of a DataFrame: from a NumPy array

 It is possible to directly create a `DataFrame` from a `NumPy` array using the `DataFrame()` constructor. The disadvantage of this method is that it is not very practical and the data type is necessarily the same for all the columns.

 Let's take a closer look at the header of this constructor.

 ``` py
 pd.DataFrame(data, index, columns, ...)
 ```
 * The `data` parameter contains the **data** to be formatted (`NumPy` array, list, dictionary or another `DataFrame`).


 * The `index` parameter, if specified, must be a **list** containing the **indices of the entries**.


 * The `columns` parameter, if specified, must be a **list** containing the **name of the columns**.
 
  <div class="alert alert-success">
<i class="fa fa-question-circle"></i> For other parameters, you can consult the Python <a href = https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html>documentation</a>.</div>
 
  <span style="color:#09b038; text-decoration : underline"> Example: </span>
 
  ``` py
  # Creation of a NumPy array with 3 rows and 4 columns
  array = np.array ([[1, 2, 3, 4],
                     [5, 6, 7, 8],
                     [9, 10, 11, 12]])
 
  # Instantiation of a DataFrame
  df = pd.DataFrame (data = array, # The data to format
                     index = ['i_1', 'i_2', 'i_3'], # The indices of each entry
                     columns = ['A', 'B', 'C', 'D']) # The name of the columns
  ```
 
  This produces the following `DataFrame`:
 
  | | A | B | C | D |
  | --------- | --- | ---- | ---- | ---- |
  |**i_1**| 1 |2| 3 |4|
  |**i_2**| 5 | 6 | 7 | 8 |
  |**i_3**|9 | 10 | 11 | 12 |
 

## 3. Creation of a DataFrame: from a dictionary
 
  Another way to create a `DataFrame` is to use a dictionary. Thanks to this technique, the columns can be of different type and their names are already given when creating the `DataFrame`.
 
  <span style="color:#09b038; text-decoration : underline"> Example: </span>
 
  ``` py
  # Creation of a dictionary
  dictionary = {'A': [1, 5, 9],
                'B': [2, 6, 10],
                'C': [3, 7, 11],
                'D': [4, 8, 12]}
 
  # Instantiation of a DataFrame
  df = pd.DataFrame (data = dictionary,
                     index = ['i_1', 'i_2', 'i_3'])
  ```
  This produces the same `DataFrame` as before:
 
  | | A | B | C | D |
 | --------- | --- | ---- | ---- | ---- |
 |**i_1**| 1 |2| 3 |4|
 |**i_2**| 5 | 6 | 7 | 8 |
 |**i_3**|9 | 10 | 11 | 12 |


The manager of a grocery store has the following stock of food products:

1. **100** jars of honey with an expiration date of **08/10/2025** and worth **€2** each.

2. **55** packets of flour expiring on **09/25/2024** each costing **€ 3**.

3. **1800** bottles of wine costing **€ 10** per unit and expiring **on 10/15/2023**. 

**(b)** From a **dictionary**, create and display a `DataFrame` **`df`** that contains **for each product**:
     * Its name.
 
 
     * Its expiration date.


    * Its quantity.
    * Its price per unit.

You will choose relevant column names and the index will be the default one (in this case we do not specify the `index` parameter).

In [5]:
food=pd.DataFrame({'name': ['Hönig', 'Mehl', 'Wein'], 'expiration_date': ['08/10/2025', '09/25/2024', '10/15/2023'],
                 'quantity': [100, 55, 1800], 'price': [2, 3, 10]})
food


name expiration_date  quantity  price
0  Hönig      08/10/2025       100      2
1   Mehl      09/25/2024        55      3
2   Wein      10/15/2023      1800     10

## 4. Creation of a DataFrame: from a data file

Most often a `DataFrame` is created directly from a file containing the data of interest. The file's format can be a CSV, Excel, txt, etc.

The most common format is the CSV format, which stands for *Comma-Separated Values* and denotes a spreadsheet-like file whose values are separated by commas.

 Here is an example:

 ```
 A, B, C, D,
 1, 2, 3, 4,
 5, 6, 7, 8,
 9, 10, 11, 12
 ```
 In this format:
 * **The first line contains the name of the columns**, but sometimes the name of the columns **is not filled in**.


 * Each **line** corresponds to an **entry** in the database.


* The values are separated by a **separator character**. In this example, it is `','` but it could be a `';'`.


 To import the data into a `DataFrame`, we need to use the `read_csv` function of `pandas` whose header is as follows:

 ```python
 pd.read_csv(filepath_or_buffer, sep = ',', header = 0, index_col = 0 ...)
 ```


 The **essential arguments** of the `pd.read_csv` function to know are:

 * **`filepath_or_buffer`**: The **path** of the .csv file relative to the execution environment. 
    * If the file is in the same folder as the Python environment, just fill in the name of the file.
    * This path must be entered in the form of **character string**.


 * **`sep`**: The character used in the .csv file to **separate** the different columns.
    * This argument must be specified as **character**.


 * **`header`**: The **number of the row** that contains the **names** of the columns.
    * If for example the column names are entered in the **first** line of the `.csv` file, then we must specify **`header = 0`**.
    * If the names are **not included**, we will put **`header = None`**.


 * **`index_col`**: The **name or number of the column** containing the **indices** of the database.
    * If the database entries are indexed by the first column, you will need to fill in **`index_col = 0`**.
    * Alternatively, if the entries are indexed by a column which bears the name `"Id"`, we can specify **`index_col = "Id"`**.


 This function will return an object of type `DataFrame` which contains all the data of the file.

* **(c)** Load the data contained in the file **`transactions.csv`** into a` DataFrame` named **`transactions`**:
 * The file is located in the **same folder** as the environment of this notebook.


 * Columns are separated by **commas**.


 * The names of the columns are in the **first line** of the file.


 *  The rows of the database are indexed by the **"transaction_id"** column which is also the **first column**.


In [6]:
import pandas as pd
transactions = pd.read_csv('transactions.csv', index_col = "transaction_id")
#transactions = pd.read_csv('transactions.csv')

print(transactions.head())



                cust_id   tran_date  prod_subcat_code  prod_cat_code  Qty  \
transaction_id                                                              
80712190438      270351  28-02-2014               1.0              1   -5   
29258453508      270384  27-02-2014               5.0              3   -5   
51750724947      273420  24-02-2014               6.0              5   -2   
93274880719      271509  24-02-2014              11.0              6   -3   
51750724947      273420  23-02-2014               6.0              5   -2   

                  Rate      Tax  total_amt Store_type  
transaction_id                                         
80712190438     -772.0  405.300  -4265.300     e-Shop  
29258453508    -1497.0  785.925  -8270.925     e-Shop  
51750724947     -791.0  166.110  -1748.110   TeleShop  
93274880719    -1363.0  429.345  -4518.345     e-Shop  
51750724947     -791.0  166.110  -1748.110   TeleShop  


We loaded the `transactions.csv` file in the` DataFrame` **`transactions`** which gathers a history of transactions carried out between 2011 and 2014. In the next section, we will study this dataset.


## 5. First exploration of a dataset using the `DataFrame` class


 The rest of this notebook briefly presents the main **methods**  of the `DataFrame` class which will allow us to do a quick analysis of our data set, that is:

 - Having a brief **overview of the data** (`head` method,` columns` and `shape` attributes).


 - **Selecting values** in the `DataFrame` (`loc` and `iloc` methods).


 - Carrying out a quick **statistical study** of our data (`describe` and `value _counts` methods)


 <div class="alert alert-info">
<i class="fa fa-info-circle"></i>
 As a reminder, to apply a method to an object in Python (such as a DataFrame for example), you must add the method as a suffix of the object. <span style="color:#09b038; text-decoration : underline"> Example: </span>my_object.my_method()</div>


## 6. Visualization of a `DataFrame`: `head` method, `columns` and `shape` attributes

 - It is possible to have a preview of a dataset by displaying **only the first lines** of the `DataFrame`.

 For that, we must use the **`head()`** method, specifying as an argument **the number of lines** that we want to display (by default 5).

 It is also possible to preview the **last lines** using the **`tail()`** method which is applied in the same way:

 ``` py
 # Display of the first 10 lines of my_dataframe
 my_dataframe.head(10)

 ```

**(d)** Display the first **20 lines** of the `transactions` `DataFrame`.

In [7]:
df_transactions = pd.DataFrame(transactions)
erste_20_zeilen = transactions.head(20)
print(erste_20_zeilen)

                cust_id   tran_date  prod_subcat_code  prod_cat_code  Qty  \
transaction_id                                                              
80712190438      270351  28-02-2014               1.0              1   -5   
29258453508      270384  27-02-2014               5.0              3   -5   
51750724947      273420  24-02-2014               6.0              5   -2   
93274880719      271509  24-02-2014              11.0              6   -3   
51750724947      273420  23-02-2014               6.0              5   -2   
97439039119      272357  23-02-2014               8.0              3   -2   
45649838090      273667  22-02-2014              11.0              6   -1   
22643667930      271489  22-02-2014              12.0              6   -1   
79792372943      275108  22-02-2014               3.0              1   -3   
50076728598      269014  21-02-2014               8.0              3   -4   
29258453508      270384  20-02-2014               5.0              3    5   

**(e)** Display the last **10 lines** of the `transactions` `DataFrame`.

In [8]:

letzte_10_zeilen = transactions.tail(10)
print(letzte_10_zeilen)

                cust_id   tran_date  prod_subcat_code  prod_cat_code  Qty  \
transaction_id                                                              
49882891062      271982  25-01-2011              10.0              5    4   
14787475597      273982  25-01-2011               4.0              3    5   
14787475597      273982  25-01-2011               4.0              3    5   
40893803228      272049  25-01-2011              11.0              6    3   
30856003613      266866  25-01-2011               4.0              2    2   
94340757522      274550  25-01-2011              12.0              5    1   
89780862956      270022  25-01-2011               4.0              1    1   
85115299378      271020  25-01-2011               2.0              6    4   
72870271171      270911  25-01-2011              11.0              5    3   
77960931771      271961  25-01-2011              11.0              5    1   

                  Rate      Tax  total_amt Store_type  
transaction_id     


  It is possible to retrieve the **name of the columns** of a `DataFrame` thanks to its **`columns`** attribute.

 ``` py
 # Creation of a DataFrame df from a dictionary
 dictionary = {'A': [1, 5, 9],
               'B': [2, 6, 10],
               'C': [3, 7, 11],
               'D': [4, 8, 12]}

 df = pd.DataFrame (data = dictionary, index = ['i_1', 'i_2', 'i_3'])
 ```
 These instructions produce the same `DataFrame` as before:

 | | A | B | C | D |
 | --------- | --- | ---- | ---- | ---- |
 | **i_ 1**| 1 | 2 | 3 | 4 |
 |**i _2** | 5 | 6 | 7 | 8 |
 | **i_ 3**| 9 | 10 | 11 | 12 |

 ``` py
 # Display of df DataFrame columns
 print(df.columns)
 >>> ['A', 'B', 'C', 'D']
 ```

 The list of the column names can be used to iterate over the columns of a `DataFrame` within a loop.

 <br>

  It can be interesting to know how many transactions (rows) and how many features (columns) the dataset contains.

 For this we will use the **`shape`** attribute of the `DataFrame` class which displays the **dimensions** of our `DataFrame` in the form of a tuple (number of rows, number of columns):
 ``` py
 # Display the dimensions of df
 print (df.shape)
 >>> (3,4)
 ```





**(f)** Display the **dimensions** of the `DataFrame` `transactions` as well as **the name of the 5th column**. Remember that in Python the indices start from 0.

In [10]:
anzahl_dimensionen = df_transactions.shape # zeigt Anzahl Zeilen und Spalten
df_transactions.columns # zeigt alle Spalten
name_5_colum= df_transactions.columns[4] # zeigt fünfte Spalte mit index 4 (0...)
print(f'Anzahl Dimensionen ist {anzahl_dimensionen}, Name vom 5. Spalte ist {name_5_colum}')


Anzahl Dimensionen ist (23053, 9), Name vom 5. Spalte ist Qty



## 7. Selecting columns from a `DataFrame`

 Extracting columns from a `DataFrame` is almost identical to extracting data from a dictionary.

 To extract a **column** from a `DataFrame`, all we have to do is enter **between brackets** the **name** of the column to extract. To extract **several** columns, we must enter between brackets **the list of the names** of the columns to extract:

 ``` py
 # Display of the 'cust_id' column
 print(transactions['cust_id'])

 # Extraction of 'cust_id' and 'Qty' columns from transactions
 cust_id_qty = transactions[["cust_id", "Qty"]]
 ```

 `cust_id_qty` is a new **` DataFrame`** containing only the `'cust_id'` and `'Qty'` columns. 

 The display of the first 3 lines of **`cust_id_qty`** yields:


 | <br><br><br> transactions_id | cust_id | Qty |
 | ----------------------------- | --------- | ----- |
 |**80712190438**| 270351 | -5 |
 |**29258453508**| 270384 | -5 |
 |**51750724947**| 273420 | -2 |


When we prepare a dataset for later use, it is better to **separate** the **categorical** variables from the **quantitative** variables:

 * A *categorical* variable is a variable that takes only a finite **number** of *modalities*.

 The categorical variables of the `DataFrame` `transactions` are: `['cust_id', 'tran_date', 'prod_subcat_code', 'prod_cat_code', 'Store_type']`.


* A *quantitative* variable is a variable that measures a quantity that can take an **infinite** number of values.

 The quantitative variables of `transactions` are: `['Qty', 'Rate', 'Tax', 'total_amt']`.


This distinction is made because some basic operations like calculating an average only make sense for quantitative variables.





* **(g)** In a `DataFrame` named **` cat_vars`**, store the **categorical** variables of `transactions`.


* **(h)** In a `DataFrame` named **` num_vars`**, store the **quantitative** variables of `transactions`.


* **(i)** Display the first 5 lines of each `DataFrame`.



In [16]:
print(df_transactions.columns) # zeigt alle Spalten

cat_vars=df_transactions[['cust_id', 'tran_date', 'prod_subcat_code', 'prod_cat_code', 'Store_type']]
num_vars=df_transactions[['Qty', 'Rate', 'Tax', 'total_amt']]
print(f'Categorical variables: {cat_vars.head(5)}\n')
print(f'Quantitative variables:{num_vars.head(5)}')

Index(['cust_id', 'tran_date', 'prod_subcat_code', 'prod_cat_code', 'Qty',
       'Rate', 'Tax', 'total_amt', 'Store_type'],
      dtype='object')
Categorical variables:                 cust_id   tran_date  prod_subcat_code  prod_cat_code  \
transaction_id                                                         
80712190438      270351  28-02-2014               1.0              1   
29258453508      270384  27-02-2014               5.0              3   
51750724947      273420  24-02-2014               6.0              5   
93274880719      271509  24-02-2014              11.0              6   
51750724947      273420  23-02-2014               6.0              5   

               Store_type  
transaction_id             
80712190438        e-Shop  
29258453508        e-Shop  
51750724947      TeleShop  
93274880719        e-Shop  
51750724947      TeleShop  

Quantitative variables:                Qty    Rate      Tax  total_amt
transaction_id                                 
807121904


## 8. Selecting rows of a `DataFrame`: ` loc` and `iloc` methods

> To extract one or more **rows** from a `DataFrame`, we use the **`loc`** method. `loc` is a very special type of method because the arguments are filled in **between square brackets** and not between parentheses. Using this method is very similar to indexing lists.
>
> In order to retrieve the line of index `i` of a `DataFrame`, all we have to do is enter `i` as an argument of the `loc` method:
>
> ``` py
> # We retrieve the line of index 80712190438 of the num_vars DataFrame
> print(num_vars.loc[80712190438])
> ```
>
> ```
>>>                 Rate    Tax  total_amt
>>> transaction_id                         
>>> 80712190438    -772.0  405.3    -4265.3
>>> 80712190438     772.0  405.3     4265.3
>
> ```
>
> In order to retrieve **several rows**, we can either:
>> * Enter a **list of indices**.
>>
>>
>> * Enter a  **slice** by specifying the start and end indices of the slice. To use **slicing** with `loc`, the indices must be **unique**, which is not the case for `transactions`.
>
> ``` py
> # We retrieve the rows at indices 80712190438, 29258453508 and 51750724947 from the transactions DataFrame
> transactions.loc[[80712190438, 29258453508, 51750724947]]
>
> ```
>
> `loc` can also take a column or **list of columns** as an argument in order to refine the data extraction:
>
> ``` python
> # We extract the columns 'Tax' and 'total_amt' from the rows at index 80712190438 and 29258453508
> transactions.loc[[80712190438, 29258453508], ['Tax', 'total_amt']]
> ```
>
> This instruction produces the following `DataFrame`:
>
> | <br><br><br> transaction_id | Tax     | total_amt |	
> |-----------------------------|---------|-----------|
> | **80712190438**             | 405.300 |	-4265.300 |
> | **80712190438**             | 405.300 |	4265.300  |
> | **29258453508**             | 785.925 |	-8270.925 |
> | **29258453508**             | 785.925 |	8270.925  |
>
>
> The **`iloc`** method is used to index a `DataFrame` **exactly like a numpy array**, that is to say by only filling in the **numeric** indices of the rows and columns. This allows the use of slicing without constraint:
>
> ``` python
> # Extraction of the first 4 rows and the first 3 columns of transactions
> transactions.iloc[0:4, 0:3]
> ```
>
> This instruction produces the following `DataFrame`:
>
>
> | <br><br><br> transaction_id | cust_id     | tran_date  | prod_subcat_code |
> |-----------------------------|-------------|------------|------------------|
> | **80712190438**             | 270351      | 28-02-2014 | 1.0              |
> | **29258453508**             | 270384      | 27-02-2014 | 5.0              |
> | **51750724947**             | 273420      | 24-02-2014 | 6.0              |
> | **93274880719**             | 271509      | 24-02-2014 | 11.0             |
>
>
> If the row indexing is the one by default (row numbering), the `loc` and `iloc` methods are **equivalent**.

## 9. Conditional indexing of a `DataFrame`

 As with Numpy arrays, we can use **conditional indexing** to extract rows from a `Dataframe` that meet a given condition.

 In the following illustration, we select the rows of the `DataFrame` `df` **for which the column `col 2` is equal to 3**.

 <br>

 <img src = "https://assets-datascientest.s3-eu-west-1.amazonaws.com/train/indexation_cond_final.png" style = 'height:200px'>

 <br>

 There are two syntaxes for conditionally indexing a `DataFrame`:

 ``` py

 # We select the rows of the DataFrame df for which the column 'col 2' is equal to 3.
 df[df['col 2'] == 3]

 df.loc[df['col 2'] == 3]

 ```

 If we want to **assign a new value** to these entries, we must absolutely use the **`loc`** method.

 Indeed, indexing with the syntax `df[df['col 2'] == 3]` only returns a **copy** of these entries and does not provide access the memory location where the data is located.

The manager of the transactions listed in the **` transactions`** `DataFrame` wishes to have access to the **identifiers** of customers who have made an **online** purchase (i.e. in a `"e-Shop"`) as well as **the date of the corresponding transaction**.

We have the following information about the columns of `transactions`:


| Column name       | Description                                          |
|:------------------|:-----------------------------------------------------|
| `'cust_id'`       | The **identifier** of the customer                 |
| `'Store_type'`    | The **type of store** where the transaction took place|
| `'tran_date'`     | The **date** of the transaction                     |

<br>

* **(j)** In a `DataFrame` named **`transactions_eshop`**, store the transactions that took place in an `"e-Shop"` type store.


* **(k)** In another `DataFrame` named **`transactions_id_date`**, store the **customer identifiers** and the **transaction date** of the `transactions_eshop` `DataFrame`.



* **(l)** Display the first 5 rows of `transactions_id_date`.



In [21]:
transactions_eshop = df_transactions.query('Store_type == "e-Shop"')
transactions_id_date = transactions_eshop[['cust_id', 'tran_date']]
transactions_id_date.head(5)

cust_id   tran_date
transaction_id                     
80712190438      270351  28-02-2014
29258453508      270384  27-02-2014
93274880719      271509  24-02-2014
45649838090      273667  22-02-2014
50076728598      269014  21-02-2014


Now, the manager would like to have access to the transactions carried out by the client whose identifier is `268819`.

* **(m)** In a `DataFrame` named **`transactions_client_268819`**, store all transactions with client identifier `268819`.


* **(n)** A column in a `DataFrame` can be iterated over with a loop exactly like a list (`for value in df['column']:`). Using a `for` loop on the `'total_amt'` column, compute and display the total transaction amount for the client with identifier `268819`.



In [30]:
transactions_client_268819 = transactions_eshop.query('cust_id == 268819')
transactions_client_268819

total = 0
for i in transactions_client_268819['total_amt']:
    total+=i
print(f'The total transaction amount ist {total:.2f} Euro')

The total transaction amount ist 14545.11 Euro


## 10. Quick statistical study of the data in a `DataFrame`.

 The **`describe`** method of a `DataFrame` returns a summary of the **descriptive statistics** (min, max, mean, quantiles,...) of its **quantitative** variables. It is therefore a very useful tool for a first visualisation of the type and distribution of these variables.

 
 
 To analyse the **categorical** variables, it is recommended to start by using the **`value_counts`** method which returns the **number of occurrences** for each modality of these variables. The `value_counts` method cannot be used directly on a `DataFrame` but only on the columns of the `DataFrame` which are objects of the **`pd.Series`** class.


* **(o)** Use the `describe` method of the `DataFrame` `transactions`. 


* **(p)** The quantitative variables of `transactions` are `'Qty'`, `'Rate'`, `'Tax'` and `total_amt'`. By default, are the statistics produced by the `describe` method  **only** computed on the quantitative variables?


* **(q)** Display the number of occurrences of each modality of the `Store_type` column using the `value_counts` method.

In [ ]:
df_transactions[['Qty', 'Rate', 'Tax', 'total_amt']]. describe() #deskriptiven Statistiken (wie Minimum, Maximum, Durchschnitt, Quantile) für die quantitativen Variablen

Qty          Rate           Tax     total_amt
count  23053.000000  23031.000000  23031.000000  23031.000000
mean       2.434173    636.405019    248.665526   2108.007267
std        2.265703    622.053592    187.087709   2506.210476
min       -5.000000  -1499.000000      7.350000  -8270.925000
25%        1.000000    312.000000     98.175000    762.450000
50%        3.000000    710.000000    199.080000   1756.950000
75%        4.000000   1109.000000    365.820000   3569.702500
max        5.000000   1500.000000    787.500000   8287.500000

In [ ]:
df_transactions. describe()#für ganze Tabelle

cust_id  prod_subcat_code  prod_cat_code           Qty  \
count   23053.000000      23021.000000   23053.000000  23053.000000   
mean   271022.241661          6.150298       3.762721      2.434173   
std      2430.830508          3.726557       1.677314      2.265703   
min    266783.000000          1.000000       1.000000     -5.000000   
25%    268936.000000          3.000000       2.000000      1.000000   
50%    270981.000000          5.000000       4.000000      3.000000   
75%    273114.000000         10.000000       5.000000      4.000000   
max    275265.000000         12.000000       6.000000      5.000000   

               Rate           Tax     total_amt  
count  23031.000000  23031.000000  23031.000000  
mean     636.405019    248.665526   2108.007267  
std      622.053592    187.087709   2506.210476  
min    -1499.000000      7.350000  -8270.925000  
25%      312.000000     98.175000    762.450000  
50%      710.000000    199.080000   1756.950000  
75%     1109.000000    365.820000   3569.702500  
max     1500.000000    787.500000   8287.500000

In [ ]:

df_transactions['Store_type'].value_counts() #die Häufigkeit jeder Kategorie in der Spalte Store_type

Store_type
e-Shop            9284
MBR               4657
Flagship store    4565
TeleShop          4493
Name: count, dtype: int64


Some transactions have **negative** amounts. 

These are transactions that have been cancelled and refunded to the client. These amounts will disrupt the distribution of the amounts which gives us **bad estimates** of the mean and quantiles of the variable `total_amt`.

* **(r)** What is the average amount of transactions with **positive** amounts?




In [39]:

positiv_transact = df_transactions[df_transactions['total_amt'] > 0]

# Durchschnitt
durchschnitt_positiv = positiv_transact['total_amt'].mean()

print(f"Durchschnittlicher Betrag positiver Transaktionen: {durchschnitt_positiv:.2f}")


Durchschnittlicher Betrag positiver Transaktionen: 2608.21
